internal/linker/link_node_modules.bzl

```diff
@@ -67,30 +58,35 @@ def write_node_modules_manifest(ctx, extra_data = [], mnemonic = None, link_work
             If source files need to be required then they can be copied to the bin_dir with copy_to_bin.
     """
 
-    mappings = {ctx.workspace_name: ["execroot", ctx.bin_dir.path]} if link_workspace_root else {}
-    node_modules_root = ""
+    mappings = {ctx.workspace_name: ["__link__", ctx.bin_dir.path]} if link_workspace_root else {}
+    node_modules_roots = {}
 
     # Look through data/deps attributes to find...
     for dep in extra_data + getattr(ctx.attr, "data", []) + getattr(ctx.attr, "deps", []):
-        # ...the root directory for the third-party node_modules; we'll symlink the local "node_modules" to it
+        # ...the root directories for the third-party node_modules; we'll symlink local "node_modules" to them
         if ExternalNpmPackageInfo in dep:
-            possible_root = "/".join([dep[ExternalNpmPackageInfo].workspace, "node_modules"])
-            if not node_modules_root:
-                node_modules_root = possible_root
-            elif node_modules_root != possible_root:
-                fail("All npm dependencies need to come from a single workspace. Found '%s' and '%s'." % (node_modules_root, possible_root))
-
-        # ...first-party packages to be linked into the node_modules tree
+            path = dep[ExternalNpmPackageInfo].path
+            workspace = dep[ExternalNpmPackageInfo].workspace
+            if path in node_modules_roots:
+                other_workspace = node_modules_roots[path]
+                if workspace != other_workspace:
+                    fail("All npm dependencies at the path '%s' must to come from a single workspace. Found '%s' and '%s'." % (path, other_workspace, workspace))
+            node_modules_roots[path] = workspace
+
+        # ...first-party packages to be linked into the root node_modules tree
         for k, v in getattr(dep, _ASPECT_RESULT_NAME, {}).items():
             if _link_mapping(dep.label, mappings, k, v):
-                # Special case for ts_library module_name for legacy behavior and for AMD name
-                # work-around. Do not propagate tslibrary root type to runtime as it is not
-                # needed at runtime. See comment below in _get_module_mappings for more info.
-                if v[0] == "_tslibrary":
-                    v = ["execroot", v[1]]
                 _debug(ctx.var, "Linking %s: %s" % (k, v))
                 mappings[k] = v
 
+    # Simplify mappings before writing them to the manifest file.
+    # The extra complication is required in bzl to handle the __tslibrary__ special
+    # case since module_name in ts_library enables the linker. In a future major release,
+    # ts_library will get a package_name attribute to enable the linker and the __tslibrary__
+    # special case can be factored out.
```

Comment:
make a 4.0 cleanup issue for this and link it here?

Reply:
done